### PR TITLE
feat: dynamic docs content structure

### PR DESCRIPTION
**Pull request by morekid**
This is the exploration for consolidating the page structure and adding the flexibility to show content based on framework.

Find the working example at `/docs/rate-limiting/quick-start`.

Things to note:
- Everything under `/docs` uses the page layout at `/src/pages/docs/[...slug].astro`, which loads a custom sidebar. Besides multiple sidebars, this also adds the flexibility of further layout customization later on.
- With this approach all frontmatter for each page will need to be generic and not dependent on the framework, and as such will be indexed for SEO. We can however customize the title/description content displayed on the page, although not always in a straightforward manner.
-  ~~All titles need to be explicitly typed in the root page if we want them to be in the automatic page TOC on the right. This is why I'm adding them as `<slots />` in a few cases. However this will always make them available for the TOC of any framework, which is also not desirable(see the multiple `4. ...` links in the page. Haven't found a workaround for this yet, but can investigate more. The fallback is that we take it into account when writing the page content: framework dependent titles won't appear in the TOC.~~
- ~~It's feasible to explore a totally custom page TOC, replacing the one that Starlight ships with.~~
- UPDATE: I've added a custom TOC component and collection schema scaffold ~~to address the above. This is a proof of concept I can refine, but it's 99% there and~~ I like that we can build the index as we like in the frontmatter for each page. ~~It's still probably a good idea to add the titles as slots as it is currently so that we can always use markdown and the `id` attrs are automatically generated~~.
- The non-docs prefixed URLs are still accessible so we might need to then redirect `/* > /docs/*`.
- The code language preference is persisted when using Starlight's `<Tabs />` built in `syncKey` prop.
- ~~This still doesn't handle linkable URLs based on framework, which in this approach should be done with params~~.

There's of course a considerable amount of manual work to port everything over the new approach.

https://github.com/user-attachments/assets/0df6f074-3302-4652-97ca-25d22122e121

### CODE REVIEW

**Comment by morekid** [comment]
cc @bendechrai

**Comment by morekid** [comment]
Added a custom TOC component approach.

**Comment by davidmytton** [comment]
Concept looks good. Why do we have to have the `/docs` prefix? Seems redundant if we're at `docs.arcjet.com` ?

At https://arcjet-docs-3jr1p35n6-arcjet.vercel.app/docs/rate-limiting/quick-start I see multiple ordered list items in the TOC:

image

**Comment by morekid** [comment]
> Concept looks good. Why do we have to have the `/docs` prefix? Seems redundant if we're at `docs.arcjet.com` ?

It is redundant but I couldn't find a quick way to fix that, perhaps we can address this with redirects (?) after all the old content is ported to the new format.

> At https://arcjet-docs-3jr1p35n6-arcjet.vercel.app/docs/rate-limiting/quick-start I see multiple ordered list items in the TOC:

Try the latest preview, hopefully you should see this:
image

**Comment by davidmytton** [comment]
Looks good - selector works nicely now.

**Comment by morekid** [comment]
@davidmytton @blaine-arcjet I went through everything I had on my list for the logic and functionality:

- Custom `ajToc` frontmatter field generates a custom TOC everywhere we need.
- The sidebar is customized as per design and the rate limiting section is a good example of how it comes out.
- I've added [nanostores](https://github.com/nanostores/nanostores) to handle state which is less verbose and (I think) has better performance than my previous approach that was just relying on local storage.
- The `<FrameworkSwitcher />` now reacts to setting `f=[framework]` in the URL so we can link to specific framework options.

There are still some visual tweaks to make this presentable but everything is ready for a test drive from you.

The content samples I've worked on are for "Get started" and "Rate limiting" at `/docs/get-started-index` and `/docs/rate-limiting/quick-start`.

**Comment by davidmytton** [comment]
This is looking great. There's a double scrollbar on the menu:

image

**Comment by morekid** [comment]
@davidmytton fixed the double scrollbar.

Not sure what the build error is, `npm run build` works fine locally.

**Comment by davidmytton** [comment]
The build error is:

```
[AstroUserError] The `labels` prop in component overrides has been removed.
  Hint:
    Replace `Astro.props.labels["Symbol(Symbol.toStringTag)"]` with `Astro.locals.t("Symbol(Symbol.toStringTag)")` instead.
    For more information see https://starlight.astro.build/guides/i18n/#using-ui-translations
  Stack trace:
    at Object.get (file:///vercel/path1/.vercel/output/static/chunks/translations_DFot5qIj.mjs:1424:13)
    at convertToSerializedForm (file:///vercel/path1/.vercel/output/static/chunks/astro/server_C0ImWB1d.mjs:211:41)
    at Array.map (<anonymous>)
    at convertToSerializedForm (file:///vercel/path1/.vercel/output/static/chunks/astro/server_C0ImWB1d.mjs:245:34)
    at Array.map (<anonymous>)
Error: Command "npm run build" exited with 1
```

**Comment by morekid** [comment]
> The build error is:
> 
> ```
> [AstroUserError] The `labels` prop in component overrides has been removed.

Yes, however we weren't explicitly using this prop anywhere. I managed to reproduce and I've fixed it with https://github.com/arcjet/arcjet-docs/pull/82/commits/2e867f88d3eed6bd48565add25c0b763e293dbed. For some reason passing all the props to the React comp was causing the break.

**Comment by morekid** [comment]
@davidmytton @blaine-arcjet still need to add basic styling to the framework select, but all is good to review functionally.

**Comment by davidmytton** [comment]
Looks good! The framework drop menu is quite subtle - I might not notice it in the top right - so I think we should have some indicator in the code snippet styling. Maybe the switcher there too? Vercel does this well:

image

**Comment by morekid** [comment]
>  I think we should have some indicator in the code snippet styling. Maybe the switcher there too?

I've added a custom component that replaces the Starlight tabs but syncs the selected tab to the same local storage field, see the video below. This is the most flexible approach I could find that also occupies minimal space.

I still have to develop it a little more to use nano stores to sync the selection live across different components + checking a console error I'm getting that might or might not be related.

The selects need proper styling still.

https://github.com/user-attachments/assets/5f85445c-a790-4c34-9dfe-72d6d751e86a

File: src/components/SelectableContent.tsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    };
+
+    // Map props to slots
+    useEffect(() => {
+      setSlots(() => {
+        return Object.entries(props)
+          .map((entry) => {
+            return { key: entry[0], value: entry[1] };
+          })
+          .sort(getComparator("desc", "key"));
```
Comment (davidmytton) [inline src/components/SelectableContent.tsx:63]: Can we make them appear in the order they are in the code? npm should show first, but also we want to show TS (App) for Next.js examples. That's the order they appear in the code, but sorting alphabetically asc or desc won't work for either of those examples.

File: src/pages/docs/[...slug].astro
Comment (davidmytton) [inline src/pages/docs/[...slug].astro:1]: @morekid Can you see if you can find a way to do the custom sidebar without needing the `/docs` prefix? The URL doesn't look good and we'll have to put in a lot of redirects.

File: src/content/docs/bot-protection/reference.mdx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+Most applications want to block almost all bots. However, it is common to allow
+some bots to access your system, such as bots for search indexing or API
+access from the command line.
+
+This behavior is configured with an `allow` list from our [full list of
+bots](https://arcjet.com/bot-list) and/or [bot
+categories](/bot-protection/identifying-bots#bot-categories).
+
+<SlotByFramework client:idle>
```
Comment (davidmytton) [inline src/content/docs/bot-protection/reference.mdx:70]: @morekid not sure what's broken here, but the code never loads - it just shows the load animation.

File: src/components/SelectableContent.tsx (hunk 1 above)
Comment (morekid) [reply src/components/SelectableContent.tsx]: @davidmytton I've improved this to index the slots (see https://github.com/arcjet/arcjet-docs/pull/82/commits/872f8be4dff085d5e8cd442940d843a51d8841eb), also updated everywhere it's used.

File: src/pages/docs/[...slug].astro
Comment (morekid) [reply src/pages/docs/[...slug].astro:1]: I couldn't find a way to make the dynamic route work with the content collection default static routes (they always [take priority](https://docs.astro.build/en/guides/routing/#route-priority-order)), so I've removed the dynamic route and custom page layout, that we can do without for the moment. However let me know if there's something I'm missing, I think we might need the custom layout flexibility eventually.

This should now all work and has no `/docs/` fragment.

File: src/content/docs/get-started.mdx [hunk 3]
```diff
@@ -3,6 +3,14 @@ title: "Get started"
 description: "Getting started with Arcjet. Quick start guide for adding rate limiting, bot protection, and email validation to your app."
 prev: false
 next: false
+pageTitle:
```
Comment (morekid) [inline src/content/docs/get-started.mdx:6]: @davidmytton page title can now be customized like this.

**Comment by morekid** [comment]
@davidmytton added plugins for expressive code for collapsing sections and showing line numbers, I've also added the Aj code syntax theme.

You can test like so on a `<Code />` component:

```
<Code
  ...
  mark={[1, {range: "3-6"}, {range: "8-12", label: "test"}]}
  ins={[{range: "13-15", label: "no"}]}
  del={[{range: "20"}]}
  collapse={"22-28"}
/>
```

**Comment by davidmytton** [comment]
Code theme looks great! It only has dark mode though - is that right? Fine by me either way.

When you collapse the viewsize, the Next.js icon doesn't show properly in the theme switcher. The others do.

image

**Comment by morekid** [comment]
> Code theme looks great! It only has dark mode though - is that right? Fine by me either way.

Yes, I don't have a light variant with expressive code atm, but everything in place and just needs some time to design/fill in the light theme vars. Regardless of whether the code stays dark in light mode it will still need brightness/contrast adjustments.

> When you collapse the viewsize, the Next.js icon doesn't show properly in the theme switcher. The others do.

Yep, will fix.

**Review by davidmytton — APPROVED** [review]
@morekid all the docs are migrated. Can you give it a final review and then it's ready to merge!

**Comment by morekid** [comment]
@davidmytton looks all good, fixed a few cosmetic issues + the Next.js icon rendering.

**Pull request merged**